self_hosted/tokenizer.jou

```diff
@@ -38,6 +37,86 @@ def is_keyword(word: byte*) -> bool:
             return True
     return False
 
+def parse_integer(string: byte*, location: Location, nbits: int) -> long:
+    if starts_with(string, "0b"):
+        base = 2
+        digits = &string[2]
+        valid_digits = "01"
+    elif starts_with(string, "0o"):
+        base = 8
+        digits = &string[2]
+        valid_digits = "01234567"
+    elif starts_with(string, "0x"):
+        base = 16
+        digits = &string[2]
+        valid_digits = "0123456789ABCDEFabcdef"
+    elif starts_with(string, "0") and strlen(string) >= 2:
+        # wrong syntax like 0777
+        fail(location, "unnecessary zero at start of number")
+        # silence compiler warnings... this code can never actually run
+        base = 0
+        digits = ""
+        valid_digits = ""
+    else:
+        # default decimal number
+        base = 10
+        digits = string
+        valid_digits = "0123456789"
+
+    if strlen(digits) == 0 or strspn(digits, valid_digits) != strlen(digits):
+        message = malloc(strlen(string) + 100)
+        sprintf(message, "invalid number or variable name \"%s\"", string)
+        fail(location, message)
+
+    # We can't use strtoll() or similar, because there's not yet a way to check errno.
+    # We would need it to check for overflow.
+    result = 0L
+    overflow = False
+    for i = 0; i < strlen(digits); i++:
+        if 'A' <= digits[i] and digits[i] <= 'F':
+            digit_value = 10 + (digits[i] - 'A')
+        elif 'a' <= digits[i] and digits[i] <= 'f':
+            digit_value = 10 + (digits[i] - 'a')
+        else:
+            digit_value = digits[i] - '0'
+
+        # Overflow isn't UB in Jou
+        if result * base < 0:
```

Comment:
This may not work correctly in bases other than 2, because an overflowing multiplication doesn't necessarily set the sign bit. But this should work:

```suggestion
        if (result * base) / base != result:
```

With any `base`, an overflowing multiplication will produce a very different value from the correct result, off by much more than `base-1`. The division cannot overflow, and it is affected by changes bigger than `base-1`.